domain/schema_validator.go

```diff
@@ -144,28 +122,34 @@ func hasRelatedTableID(relatedTableIDs, updateTableIDs []int64) bool {
 // from usedVer to the latest schema version.
 // NOTE, this function should be called under lock!
 func (s *schemaValidator) isRelatedTablesChanged(currVer int64, tableIDs []int64) bool {
-	if _, ok := s.itemSchemaVers[currVer]; !ok {
+	if len(s.detalItemInfos) == 0 {
+		log.Infof("schema change history is empty, checking %d", currVer)
+		return true
+	}
+	oldestVersion := s.detalItemInfos[0].schemaVersion
+	if versionNewerThan(oldestVersion, currVer) {
 		log.Infof("the schema version %d is much older than the latest version %d", currVer, s.latestSchemaVer)
 		return true
 	}
 
-	// Find currVer's offset in detaItemInfos.
-	offset := 0
-	for i, info := range s.detalItemInfos {
-		if info.schemaVersion == currVer {
-			offset = i
+	for i := len(s.detalItemInfos) - 1; i > 0; i-- {
+		item := &s.detalItemInfos[i]
+		if !versionNewerThan(item.schemaVersion, currVer) {
```

Comment:
How about
```
if currVer > item.schemaVersion {
    return false
}
```

Then in the end, return `true`?

Then the line from 129 to 133 can be removed.

Reply:
Actually,  line 129~133 is an optimization to avoid scan the whole queue, in some case.

Reply:
Just the same with the old logic:

```
if _, ok := s.itemSchemaVers[currVer]; !ok {
    log.Infof("the schema version %d is much older than the latest version %d", currVer, s.latestSchemaVer)
    return true
}
```